### PR TITLE
refactor: add typescript

### PR DESCRIPTION
**Pull request by fjlanasa**
#### Summary of changes
**Asana Ticket:** [💳  add typescript to signs-ui](TICKET_LINK)

Changes:
* remove babel dependencies and configuration file
* add typescript, type declarations for dependencies, and necessary webpack plugins (along with webpack config changes)
* update `eslint` config to work with typescript and without babel
* rename `.js` -> `.ts` and `.jsx` -> `.tsx`. Unfortunately, I couldn't make git recognize the latter as a rename, so they appear as new files here
* move test files into separate `test` directory to make it easier to avoid loading with `ts-loader`
* remove `prop-types` library and replaced its usage with typescript interfaces
* various small updates to source files to pass typescript linter

#### Reviewer Checklist
- [ ] Meets ticket's acceptance criteria
- [ ] Any new or changed functions have typespecs
- [ ] Tests were added for any new functionality (don't just rely on Codecov)
- [ ] This branch was deployed to the staging environment and is currently running with no unexpected increase in warnings, and no errors or crashes (compare on Splunk: [staging](https://mbta.splunkcloud.com/en-US/app/search/search?q=search%20index%3Dsigns-ui-dev%20%22%5Berror%5D%22%20OR%20%22%5Bwarn%5D%22%20OR%20%22CRASH%22&display.page.search.mode=verbose&dispatch.sample_ratio=1&earliest=-4h%40m&latest=now&sid=1545840769.3874970) vs. [prod](https://mbta.splunkcloud.com/en-US/app/search/search?q=search%20index%3Dsigns-ui-prod%20%22%5Berror%5D%22%20OR%20%22%5Bwarn%5D%22%20OR%20%22CRASH%22&display.page.search.mode=verbose&dispatch.sample_ratio=1&earliest=-4h%40m&latest=now&sid=1545840745.3874956))

### CODE REVIEW

File: assets/js/types.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      n: ZoneConfig;
+      e: ZoneConfig;
+      s: ZoneConfig;
+      w: ZoneConfig;
+      c: ZoneConfig;
+      m: ZoneConfig;
+    };
+  }
+
+type SignConfig = { id?: string, mode: 'auto' | 'headway' | 'off' | 'static_text', expires?: string | null, line1?: string; line2?: string}
```
Comment (digitalcora) [inline assets/js/types.ts:33]: This seems quite long. Do we have a line length limit as part of our code prettification/format checking?

File: assets/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        // "outDir": "./",                     /* Redirect output structure to the directory. */
+        // "rootDir": "./",                    /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
+        // "composite": true,                  /* Enable project compilation */
+        // "removeComments": true,             /* Do not emit comments to output. */
+        // "noEmit": true,                     /* Do not emit outputs. */
+        // "importHelpers": true,              /* Import emit helpers from 'tslib'. */
+        // "downlevelIteration": true,         /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
+        // "isolatedModules": true,            /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */
+        /* Strict Type-Checking Options */
+        // "strict": true, /* Enable all strict type-checking options. */
```
Comment (digitalcora) [inline assets/tsconfig.json:26]: Is it practical for us to enable full `strict` mode instead of just the `strictNullChecks`?

File: assets/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        // "strictFunctionTypes": true, /* Enable strict checking of function types. */
+        // "strictBindCallApply": true, /* Enable strict 'bind', 'call', and 'apply' methods on functions. */
+        // "strictPropertyInitialization": true, /* Enable strict checking of property initialization in classes. */
+        // "noImplicitThis": true, /* Raise error on 'this' expressions with an implied 'any' type. */
+        // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */
+        // /* Additional Checks */
+        // "noUnusedLocals": true, /* Report errors on unused locals. */
+        // "noUnusedParameters": true, /* Report errors on unused parameters. */
+        // "noImplicitReturns": true, /* Report error when not all code paths in function return a value. */
+        // "noFallthroughCasesInSwitch": true, /* Report errors for fallthrough cases in switch statement. */
```
Comment (digitalcora) [inline assets/tsconfig.json:38]: Should we enable any of these Additional Checks or are they already covered by the linting?

**Review by digitalcora — COMMENTED** [review]
Nice work taking this on!!

File: assets/js/Line.tsx [hunk 4]
```diff
@@ -64,11 +64,11 @@ function Line({
   readOnly,
   configuredHeadways,
   setConfiguredHeadways,
-  chelseaBridgeAnnouncements,
-  setChelseaBridgeAnnouncements,
```
[the remark this answers is not in the document]
Comment (fjlanasa) [reply assets/js/Line.tsx:68]: oh, yep. good catch.

File: assets/js/Line.tsx [hunk 5]
```diff
@@ -204,33 +187,20 @@ function Line({
...
-    })),
-  })),
-};
-
 Line.defaultProps = {
-  stationConfigs: null,
+  stationConfigs: undefined,
 };
 
+interface LineProps {
```
[the remark this answers is not in the document]
Comment (fjlanasa) [reply assets/js/Line.tsx:194]: yeah, definitely. forgot to move it up there after copying over the attributes from `prop-types`.

File: assets/js/types.ts (hunk 1 above)
Comment (fjlanasa) [reply assets/js/types.ts:33]: i actually don't think we do. we haven't actually set up anything for formatting besides running `eslint --fix`, which will do things like change double quotes into single quotes, but apparently not this. I thought about adding that onto this PR, but was trying to limit the diff.

File: assets/tsconfig.json (hunk 2 above)
Comment (fjlanasa) [reply assets/tsconfig.json:26]: i can check

File: assets/tsconfig.json (hunk 3 above)
Comment (fjlanasa) [reply assets/tsconfig.json:38]: hmmm. i'll give it a little more thought. i just copied this over from 🏹

File: assets/js/types.ts (hunk 1 above)
Comment (digitalcora) [reply assets/js/types.ts:33]: hm, I see. created an inbound task for us to think about: https://app.asana.com/0/881264583703207/1199158942990536

File: assets/js/Line.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (digitalcora) [reply assets/js/Line.tsx:194]: If we want to do this consistently I imagine we could remove the `"no-use-before-define": 0` from the ESLint config and let it catch all instances (unless this is referring to something else).

File: assets/js/Line.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (fjlanasa) [reply assets/js/Line.tsx:68]: i accidentally included adding this code back in this commit, which added more typecheck rules https://github.com/mbta/signs_ui/pull/381/commits/250a96a39f77ed7ee534ab6d4a2c12ad8fe1c18d

File: assets/js/Line.tsx [hunk 6]
```diff
@@ -97,26 +118,28 @@ function Line({
...
-          />
-          <span className="switch-label">Switch</span>
-        </div>
-      </label>
-      )}
+
       {!readOnly && (
-        <div className="viewer--toggle-all">
-          {batchModes.auto && (
+        <>
```
Comment (digitalcora) [inline assets/js/Line.tsx:123]: It looks like there's an extra change here to move the bridge announcements switch inside the `!readOnly` block — I guess this is something [we want to fix](https://app.asana.com/0/881264583703207/1199112192455310) anyway, but it should probably be its own commit, if not its own PR (and have tests and stuff).

Comment (fjlanasa) [reply assets/js/Line.tsx:123]: ok i'll remove it then

**Review by digitalcora — APPROVED** [review]
Pending the extraction of the drawbridge read-only fix, I say push onward before this accumulates any more conflicts!

**Comment by fjlanasa** [comment]
> Pending the extraction of the drawbridge read-only fix, I say push onward before this accumulates any more conflicts!

I'm not extracting. I'm just not doing it here and we can take care of it in the existing ticket.
